server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/VariableWidthHistogramAggregator.java

```diff
@@ -0,0 +1,378 @@
+
+        bigArrays = context.bigArrays();
+        this.cachedValues = bigArrays.newDoubleArray(0);
+        this.clusterMins = bigArrays.newDoubleArray(0);
+        this.clusterMaxes = bigArrays.newDoubleArray(0);
+        this.clusterCentroids = bigArrays.newDoubleArray(0);
+        this.clusterSizes = bigArrays.newIntArray(0);
+
+        String scoringAgg = subAggsNeedScore();
+        String nestedAgg = descendsFromNestedAggregator(parent);
+        if (scoringAgg != null && nestedAgg != null) {
+            /*
+             * Terms agg would force the collect mode to depth_first here, because
+             * we need to access the score of nested documents in a sub-aggregation
+             * and we are not able to generate this score while replaying deferred documents.
+             *
+             * But the VariableWidthHistogram agg _must_ execute in breadth first since it relies on
+             * deferring execution, so we just have to throw up our hands and refuse
+            */
+            throw new IllegalStateException("VariableWidthHistogram agg [" + name() + "] is the child of the nested agg [" + nestedAgg
+                + "], and also has a scoring child agg [" + scoringAgg + "].  This combination is not supported because " +
+                "it requires executing in [depth_first] mode, which the VariableWidthHistogram agg cannot do.");
+        }
+    }
+
+    private String subAggsNeedScore() {
+        for (Aggregator subAgg : subAggregators) {
+            if (subAgg.scoreMode().needsScores()) {
+                return subAgg.name();
+            }
+        }
+        return null;
+    }
+
+    private String descendsFromNestedAggregator(Aggregator parent) {
+        while (parent != null) {
+            if (parent.getClass() == NestedAggregator.class) {
+                return parent.name();
+            }
+            parent = parent.parent();
+        }
+        return null;
+    }
+
+    @Override
+    public ScoreMode scoreMode() {
+        if (valuesSource != null && valuesSource.needsScores()) {
+            return ScoreMode.COMPLETE;
+        }
+        return super.scoreMode();
+    }
+
+    @Override
+    protected boolean shouldDefer(Aggregator aggregator) {
+        return true;
+    }
+
+    @Override
+    public DeferringBucketCollector getDeferringCollector() {
+        deferringCollector = new MergingBucketsDeferringCollector(context);
+        return deferringCollector;
+    }
+
+    @Override
+    protected LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCollector sub) throws IOException {
+        if (valuesSource == null) {
+            return LeafBucketCollector.NO_OP_COLLECTOR;
+        }
+        final SortedNumericDoubleValues values = valuesSource.doubleValues(ctx);
+        return new LeafBucketCollectorBase(sub, values){
+            @Override
+            public void collect(int doc, long bucket) throws IOException {
+                assert bucket == 0;
```

Comment:
We've recently deprecated `asMultiBucketAggregator` which is the reason you can assert this because it has some fairly nasty side effects from a Lucene standpoint. OTOH, reworking this so it supports multiple `owningBucketOrd`s would be a pretty big pain. So I'm super OK merging without doing that and coming back to it later.

Reply:
As I made this PR about a year ago, it would be nice to pull new changes like this one into my branch. 

However, I'm not sure if merging master into the branch would be a good idea: when I've done this in the past, GitHub put *all* of the new merged changes into my PR diffs. I'm not sure if that would happen here. But if it did, this would make the PR really hard to read.

`CONTRIBUTING.md` says not to rebase without permission. Which option would you recommend here?

Reply:
A year is a long time, yeah. If you are up for rebasing that'd be fine, though you probably would want to squash first too. In general that makes reviewing a lot harder because I can't easily look at the "diff" that you made. But I'm I wouldn't mind being forced to read the whole thing again. So, yeah +1 to squash it to one, rebase, and force push to your branch.

Reply:
Ok sounds good. In that case, I will just do that at the very end so that reviewing is easier for you.